// color-core/dist/harmony/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.triadic = exports.tones = exports.tints = exports.tetradic = exports.square = exports.splitComplementary = exports.shades = exports.monochromatic = exports.doubleSplitComplementary = exports.complementary = exports.analogous = void 0;
var analogous_1 = require("./components/analogous");
Object.defineProperty(exports, "analogous", { enumerable: true, get: function () { return analogous_1.analogous; } });
var complementary_1 = require("./components/complementary");
Object.defineProperty(exports, "complementary", { enumerable: true, get: function () { return complementary_1.complementary; } });
var double_split_complementary_1 = require("./components/double-split-complementary");
Object.defineProperty(exports, "doubleSplitComplementary", { enumerable: true, get: function () { return double_split_complementary_1.doubleSplitComplementary; } });
var monochromatic_1 = require("./components/monochromatic");
Object.defineProperty(exports, "monochromatic", { enumerable: true, get: function () { return monochromatic_1.monochromatic; } });
var shades_1 = require("./components/shades");
Object.defineProperty(exports, "shades", { enumerable: true, get: function () { return shades_1.shades; } });
var split_complementary_1 = require("./components/split-complementary");
Object.defineProperty(exports, "splitComplementary", { enumerable: true, get: function () { return split_complementary_1.splitComplementary; } });
var square_1 = require("./components/square");
Object.defineProperty(exports, "square", { enumerable: true, get: function () { return square_1.square; } });
var tetradic_1 = require("./components/tetradic");
Object.defineProperty(exports, "tetradic", { enumerable: true, get: function () { return tetradic_1.tetradic; } });
var tints_1 = require("./components/tints");
Object.defineProperty(exports, "tints", { enumerable: true, get: function () { return tints_1.tints; } });
var tones_1 = require("./components/tones");
Object.defineProperty(exports, "tones", { enumerable: true, get: function () { return tones_1.tones; } });
var triadic_1 = require("./components/triadic");
Object.defineProperty(exports, "triadic", { enumerable: true, get: function () { return triadic_1.triadic; } });
