// color-core/dist/manipulation/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adjustSaturation = exports.mix = exports.adjustLightness = exports.invert = exports.adjustHue = exports.grayscale = exports.adjustAlpha = void 0;
var alpha_1 = require("./components/alpha");
Object.defineProperty(exports, "adjustAlpha", { enumerable: true, get: function () { return alpha_1.adjustAlpha; } });
var grayscale_1 = require("./components/grayscale");
Object.defineProperty(exports, "grayscale", { enumerable: true, get: function () { return grayscale_1.grayscale; } });
var hue_1 = require("./components/hue");
Object.defineProperty(exports, "adjustHue", { enumerable: true, get: function () { return hue_1.adjustHue; } });
var invert_1 = require("./components/invert");
Object.defineProperty(exports, "invert", { enumerable: true, get: function () { return invert_1.invert; } });
var lightness_1 = require("./components/lightness");
Object.defineProperty(exports, "adjustLightness", { enumerable: true, get: function () { return lightness_1.adjustLightness; } });
var mix_1 = require("./components/mix");
Object.defineProperty(exports, "mix", { enumerable: true, get: function () { return mix_1.mix; } });
var saturation_1 = require("./components/saturation");
Object.defineProperty(exports, "adjustSaturation", { enumerable: true, get: function () { return saturation_1.adjustSaturation; } });
