// color-core/dist/conversions/components/hpluv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hpluvToRgb = hpluvToRgb;
exports.rgbToHPLuv = rgbToHPLuv;
exports.hpluvToLch = hpluvToLch;
exports.lchToHpluv = lchToHpluv;
var helpers_1 = require("./helpers");
var xyz_1 = require("./xyz");
/**
 * Rounds an RGB value to the nearest integer and clamps it between 0 and 255.
 * @param value - The RGB value to round and clamp.
 * @returns The rounded and clamped RGB value.
 */
function roundAndClampRgb(value) {
    return Math.max(0, Math.min(255, Math.round(value)));
}
/**
 * Converts HPLuv color to RGB.
 * @param hpluv - The HPLuv color to convert.
 * @returns The converted RGB color.
 */
function hpluvToRgb(hpluv) {
    // Handle edge cases
    if (hpluv.l === 0)
        return { r: 0, g: 0, b: 0 };
    if (hpluv.l === 100)
        return { r: 255, g: 255, b: 255 };
    var lch = hpluvToLch(hpluv);
    var luv = (0, helpers_1.lchToLuv)(lch);
    var xyz = (0, helpers_1.luvToXyz)(luv);
    var rgb = (0, xyz_1.xyzToRgb)(xyz);
    // Round and clamp RGB values
    return {
        r: roundAndClampRgb(rgb.r),
        g: roundAndClampRgb(rgb.g),
        b: roundAndClampRgb(rgb.b)
    };
}
/**
 * Converts RGB color to HPLuv.
 * @param rgb - The RGB color to convert.
 * @returns The converted HPLuv color.
 */
function rgbToHPLuv(rgb) {
    // Handle edge cases
    if (rgb.r === 0 && rgb.g === 0 && rgb.b === 0)
        return { h: 0, p: 0, l: 0 };
    if (rgb.r === 255 && rgb.g === 255 && rgb.b === 255)
        return { h: 0, p: 0, l: 100 };
    var xyz = (0, xyz_1.rgbToXyz)(rgb);
    var luv = (0, helpers_1.xyzToLuv)(xyz);
    var lch = (0, helpers_1.luvToLch)(luv);
    return lchToHpluv(lch);
}
/**
 * Converts HPLuv to LCH.
 * @param hpluv - The HPLuv color to convert.
 * @returns The converted LCH color.
 */
function hpluvToLch(hpluv) {
    if (hpluv.l > 99.9999999) {
        return { l: 100, c: 0, h: hpluv.h };
    }
    if (hpluv.l < 0.00000001) {
        return { l: 0, c: 0, h: hpluv.h };
    }
    var l = hpluv.l;
    var lines = (0, helpers_1.calculateBoundingLines)(l);
    var max = (0, helpers_1.calcMaxChromaHpluv)(lines);
    var c = max / 100 * hpluv.p;
    return { l: l, c: c, h: hpluv.h };
}
/**
 * Converts LCH to HPLuv.
 * @param lch - The LCH color to convert.
 * @returns The converted HPLuv color.
 */
function lchToHpluv(lch) {
    if (lch.l > 99.9999999) {
        return { h: lch.h, p: 0, l: 100 };
    }
    if (lch.l < 0.00000001) {
        return { h: lch.h, p: 0, l: 0 };
    }
    var lines = (0, helpers_1.calculateBoundingLines)(lch.l);
    var max = (0, helpers_1.calcMaxChromaHpluv)(lines);
    var p = lch.c / max * 100;
    return { h: lch.h, p: p, l: lch.l };
}
